### PR TITLE
feat!: support for custom proposal-based voting periods 

### PR DESCRIPTION
**Pull request by alexanderbez**
Add support for custom voting periods:

- Define `ProposalVotingPeriod`
- Add `[]ProposalVotingPeriod` to the existing `VotingParams` (on-chain upgrade should be fine)
- Update `Keeper#ActivateVotingPeriod` to account for the new `[]ProposalVotingPeriod`

**NOTE**: I did NOT add/implement `CustomVotingPeriodQuorum` as I just didn't see that as necessary.

ref: https://github.com/osmosis-labs/osmosis/issues/1517

### CODE REVIEW

**Comment by alexanderbez** [comment]
Will fix tests shortly.

File: x/gov/keeper/grpc_query.go [hunk 1]
```diff
@@ -169,8 +169,8 @@ func (q Keeper) Params(c context.Context, req *types.QueryParamsRequest) (*types
 		return &types.QueryParamsResponse{DepositParams: depositParmas}, nil
 
 	case types.ParamVoting:
-		votingParmas := q.GetVotingParams(ctx)
-		return &types.QueryParamsResponse{VotingParams: votingParmas}, nil
+		votingParams := q.GetVotingParams(ctx)
+		return &types.QueryParamsResponse{VotingParams: votingParams}, nil
```
Comment (mattverse) [inline x/gov/keeper/grpc_query.go:173]: whoa didnt know we had a typo all this time :p

**Review by mattverse — APPROVED** [review]
LGTM! 

Nit(we can also create an issue and set this as a chore): I wish there was more test cases added, other than simply comparing the two proposals, but don't want to block merging for this!

File: x/gov/keeper/grpc_query.go (hunk 1 above)
Comment (alexanderbez) [reply x/gov/keeper/grpc_query.go:173]: lol...I'm sure more exist :p

**Comment by alexanderbez** [comment]
> LGTM!
> 
> Nit(we can also create an issue and set this as a chore): I wish there was more test cases added, other than simply comparing the two proposals, but don't want to block merging for this!

@mattverse I added `TestGetVotingPeriod`, which is the core change in this PR (apart from adding the new types). That should be enough to cover things.

File: x/gov/keeper/grpc_query.go (hunk 1 above)
Comment (ValarDragon) [reply x/gov/keeper/grpc_query.go:173]: lmao, imagine if this got further auto-completed to `parmesians`

**Review by ValarDragon — APPROVED** [review]
Awesome, looks good to me!

We shouldn't need to initialize the new param in our app.go right, since this doesn't update the paramset, just the concrete struct at `VotingParams`?

**Comment by alexanderbez** [comment]
> Awesome, looks good to me!
> 
> We shouldn't need to initialize the new param in our app.go right, since this doesn't update the paramset, just the concrete struct at `VotingParams`?

That's correct!

**Comment by assafmo** [comment]
This is super cool guys!

**Pull request merged**